### PR TITLE
Basic tests added to the findFirst function

### PR DESCRIPTION
**Pull request by jpsullivan**
This isn't ready to be merged yet. I have some questions about `findFirst` usage. Testing against something like this is a no brainer:
```javascript
hi("a string to be searched").findFirst("to be");
```

My question for you is whether the following is something `findFirst` should be able to evaluate:
```javascript
hi("this shouldn't find anything", "this should though").findFirst("though");

... or ...

hi(["a", "string", "to", "be", "searched"]).findFirst("string");
```
Each returns `undefined` for me. Is that the expected result or have I just found a bug? I see code here to handle search subjects of n-elements, but the loop only works for multiple string arguments, which just loops over each character in a string. Arrays never get evaluated.

If this is a bug I'll go on and fix it. Just wanted to see if I'm somehow using this completely wrong.

### CODE REVIEW

File: src/functions/findFirst.js [hunk 1]
```diff
@@ -7,12 +7,27 @@ import {copyable} from "./copyable";
...
-// common use case for substring searching, so alias "find" to "findFirst"
-// for maximum user convenience and minimum user confusion.
 export const findFirst = wrap({
     names: ["findFirst", "find"],
+    summary: "Finds the first occurrance of a substring as judged by a comparison function.",
+    docs: process.env.NODE_ENV !== "development" ? undefined : {
+        introduced: "higher@1.0.0",
+        detail: (`
+
+        `),
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: If there's no detail docstring, the detail attribute ought to be absent

File: src/functions/findFirst.js [hunk 2]
```diff
@@ -7,12 +7,27 @@ import {copyable} from "./copyable";
...
+        expects: (`
+            The function expects as input a comparison function and a sequence array
+            which should contain a source sequence and a bounded search sequence.
+
+            The function expects as input a sequence known to be bounded and
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: To start a new paragraph precede the line with a "/"

But in general avoid blank lines, imo it makes it less readable but we could discuss that if you feel strongly otherwise

File: src/functions/findFirst.js [hunk 3]
```diff
@@ -7,12 +7,27 @@ import {copyable} from "./copyable";
...
+            an optional predicate function to apply to each element.
+        `),
+        returns: (`
+            The function returns a @FindSequenceResult if an occurrance of the
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: "occurrence" not "occurrance"

File: src/functions/findFirst.js [hunk 4]
```diff
@@ -7,12 +7,27 @@ import {copyable} from "./copyable";
...
+            provided substring can be found. If none can be found, or the sequence
+            is empty, returns @undefined.
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: Hmm `@undefined` is fine but I have yet to really decide whether it's better to indicate literals as that or with backticks

```
`undefined`
```

what do you think?

**Comment by pineapplemachine** [comment]
Here are some examples of tests from my similar D `find` implementation

These are for `findAll` but in this `findAll` and `findFirst`  and `findLast` use the same implementation with different template arguments

``` js
tests("Overlapping", {
    auto result = allfunc!eq("etetet", "etet");
    testeq(result.length, 2);
    testeq(result[0].index, 0);
    testeq(result[0].value, "etet");
    testeq(result[1].index, 2);
    testeq(result[1].value, "etet");
});
```

``` js
tests("Single-length subject", {
    auto result = allfunc!eq("abcabc", "a");
    testeq(result.length, 2);
    testeq(result[0].index, 0);
    testeq(result[0].value, "a");
    testeq(result[1].index, 3);
    testeq(result[1].value, "a");
});
```

``` js
tests("Empty source", {
    auto range = "".findalliterlazy(transformsubject("hi"));
    test(range.empty);
});
```

``` js
tests("Empty Subject", {
    testfail({
        auto range = "test".findalliterlazy(transformsubject(""));
    });
});
```

File: src/functions/findFirst.js [hunk 5]
```diff
@@ -78,6 +93,32 @@ export const findFirst = wrap({
         }
         return undefined;
     },
+    tests: process.env.NODE_ENV !== "development" ? undefined : {
+        "basicUsage": (hi) => {
+            const seq = hi("a string to be searched").findFirst("to be");
+            hi.assert(seq, (r) => r.index === 9);
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: This assert isn't correct

The first arg is a condition evaluated for truthiness and the second is an (optional) error message for when it fails

Comment (pineapplemachine) [reply src/functions/findFirst.js]: Also for a basic usage example maybe something like

``` js
hi.assert(hi("hello world").findFirst("world").index === 6);
hi.assertUndefined(hi("hello again").findFirst("nope"));
```

Comment (pineapplemachine) [reply src/functions/findFirst.js]: There should also be a similar and separate example for case-insensitive substring search e.g.

``` js
const asciiCaseInsensitive = (a, b) => (a.toUpperCase() === b.toUpperCase());
const found = hi("hello world").findFirst("World", asciiCaseInsensitive);
hi.assert(found.index === 6);
hi.assert(found.string() === "world");
```

File: src/functions/findFirst.js [hunk 6]
```diff
@@ -7,12 +7,27 @@ import {copyable} from "./copyable";
 import {equals} from "./equals";
 import {FindSequenceResult, ForwardFindSequenceThread, stepFindThreads} from "./findAll";
 
-// Find the first occurrence of a substring as judged by a comparison function.
-// Finding the first instance of a substring is overwhelmingly the most
-// common use case for substring searching, so alias "find" to "findFirst"
-// for maximum user convenience and minimum user confusion.
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: ...I would actually like to keep an explanation for the alias

File: src/functions/findFirst.js [hunk 7]
```diff
@@ -7,12 +7,27 @@ import {copyable} from "./copyable";
...
+            which should contain a source sequence and a bounded search sequence.
+
+            The function expects as input a sequence known to be bounded and
+            an optional predicate function to apply to each element.
+        `),
+        returns: (`
+            The function returns a @FindSequenceResult if an occurrance of the
+            provided substring can be found. If none can be found, or the sequence
+            is empty, returns @undefined.
+        `),
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: Does it throw an error when the input sequence wasn't bounded? (It should) and there should be a `throws` docstring (see a function e.g. `any` for reference)

**Comment by pineapplemachine** [comment]
Needs test cases for an unbounded input sequence, empty input sequence, empty search string, sequences other than character strings

**Comment by pineapplemachine** [comment]
shit I hit the wrong button I didn't meant to close sorry

**Comment by jpsullivan** [comment]
Thanks for the great review! Will fix.

File: src/functions/findFirst.js (hunk 2 above)
Comment (jpsullivan) [inline src/functions/findFirst.js]: Yeah I was going to eventually put something in here. Just emptied it out to remember that I hadn't done it yet :)

File: src/functions/findFirst.js (hunk 3 above)
Comment (jpsullivan) [reply src/functions/findFirst.js]: Doh!

File: src/functions/findFirst.js (hunk 5 above)
Comment (jpsullivan) [reply src/functions/findFirst.js]: Great ideas here. I'm awful at the basic usage tests lol. Tests as documentation is tricky for me, which is probably a good thing.

File: src/functions/findFirst.js (hunk 4 above)
Comment (jpsullivan) [reply src/functions/findFirst.js]: I think having it with backticks probably makes more sense since it's used mainly as a return type.

File: src/functions/findFirst.js (hunk 1 above)
Comment (jpsullivan) [reply src/functions/findFirst.js]: Yeah I was going to eventually put something in here. Just emptied it out to remember that I hadn't done it yet :)

**Comment by jpsullivan** [comment]
This is unfortunately still incomplete. Been short on time lately. Should have this all finished within the next 24 hours though. Sorry!

File: src/functions/findFirst.js [hunk 8]
```diff
@@ -117,6 +120,22 @@ export const findFirst = wrap({
         "searchLengthCorrect": (hi) => {
             const result = hi("a string to be searched").findFirst("string");
             hi.assert(result, (r) => r.length === 6);
+        },
+        "caseSensitiveSearch": (hi) => {
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: Case INsensitive!

Comment (pineapplemachine) [reply src/functions/findFirst.js]: Also can we add this one to `examples` along with `basicUsage`?

**Comment by pineapplemachine** [comment]
Updated tests are looking pretty good

File: src/functions/findFirst.js [hunk 9]
```diff
@@ -136,6 +147,22 @@ export const findFirst = wrap({
...
             hi.assert(result.high === 4);
+        },
+        "emptySearchString": (hi) => {
+            hi.assertFailWith(NotBoundedError, () => {
+                return hi("test").findFirst("");
+            });
+        },
+        "nonStringSearchSequence": (hi) => {
+            hi.assertFailWith(NotBoundedError, () => {
+                return hi(true, false, {}).findFirst({});
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: `hi` accepts a single argument and the rest are ignored

File: src/functions/findFirst.js [hunk 10]
```diff
@@ -136,6 +147,22 @@ export const findFirst = wrap({
...
+            });
+            hi.assertUndefined(NotBoundedError, () => {
+                return hi(true, false, {}).findFirst("true");
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: If the `hi` call actually produced a sequence this would be legal (currently it should be returning `undefined` because `true` is not valid as a sequence)

**Comment by pineapplemachine** [comment]
also - I'm not sure that it's appropriate to throw a `NotBoundedError` when the search string was empty

File: src/functions/findFirst.js [hunk 11]
```diff
@@ -48,12 +48,18 @@ export const findFirst = wrap({
...
-        if(search.done() || search.unbounded()){
+        // simply return undefined if the search input is empty
+        if (search.done()){
+            return undefined;
+        }
+
+        // ... however throw a NotBoundedError if the search input is unbounded
+        if(search.unbounded()){
             throw NotBoundedError(search, {
                 message: "Search sequence is either empty or unbounded"
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: Update the error message - ought to read "Failed to find substring", the error constructor adds to the message an explanation as to why. (the sequence was unbounded, was not known to be bounded, whatever)

File: src/functions/findFirst.js [hunk 12]
```diff
@@ -115,15 +121,10 @@ export const findFirst = wrap({
             const seq = hi("not gonna find this").findFirst("hello?");
             hi.assertUndefined(seq);
         },
-        "undefinedWhenSearchIsEmpty": (hi) => {
+        "undefinedWhenSequenceIsEmpty": (hi) => {
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: duplicate of `emptySearchString` test

File: src/functions/findFirst.js [hunk 13]
```diff
@@ -149,20 +150,19 @@ export const findFirst = wrap({
...
         "emptySearchString": (hi) => {
+            hi.assertUndefined(hi("test").findFirst(""));
+        },
+        "unboundedSearchInput": (hi) => {
             hi.assertFailWith(NotBoundedError, () => {
-                return hi("test").findFirst("");
+                return hi("test").findFirst(hi.repeatElement({
+                    key: "hello",
+                    value: "world"
+                }));
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: Can we just like, `hi.repeatElement("x")` instead? This seems unnecessarily hard on the reader trying to understand what the test does

File: src/functions/findFirst.js [hunk 14]
```diff
@@ -149,20 +150,19 @@ export const findFirst = wrap({
...
             });
         },
         "nonStringSearchSequence": (hi) => {
```
Comment (pineapplemachine) [inline src/functions/findFirst.js]: This function is valid for substrings other than literal string literals. Rename test case to `emptySearchSequence` perhaps? Add also a blank string input to go with the others?

**Comment by pineapplemachine** [comment]
Closing per discussion in DMs, to be integrated into https://github.com/pineapplemachine/higher/issues/99 which will affect the behavior of this function. Most or all of these tests should remain valid, however, so this branch will be used as a starting point for that task.
